lightcrafts/src/main/java/com/lightcrafts/model/ImageEditor/ZoneFinder.java

```diff
@@ -316,52 +392,86 @@ private RenderedImage segment(RenderedImage image) {
     class Segmenter extends Thread {
-        RenderedImage image;
-        RenderedImage nextImage = null;
+        private final BlockingQueue<SegmentRequest> requestQueue = new LinkedBlockingQueue<>(1);
+        private final AtomicBoolean running = new AtomicBoolean(true);
 
         Segmenter(Rectangle visibleRect, PlanarImage image) {
-            super("ZoneFinder Histogrammer");
-            this.image = cropScaleGrayscale(visibleRect, image);
+            super("ZoneFinder Segmenter");
+            setDaemon(true);
+            requestQueue.offer(new SegmentRequest(visibleRect, image));
         }
 
-        synchronized void nextView(Rectangle visibleRect, PlanarImage image) {
-            nextImage = cropScaleGrayscale(visibleRect, image);
+        void nextView(Rectangle visibleRect, PlanarImage image) {
+            // Clear old requests and add new one
+            requestQueue.clear();
+            requestQueue.offer(new SegmentRequest(visibleRect, image));
         }
 
-        synchronized private boolean getNextView() {
-            if (nextImage != null) {
-                image = nextImage;
-                nextImage = null;
-                return true;
-            } else
-                return false;
+        void shutdown() {
+            running.set(false);
+            interrupt();
         }
 
         @Override
         public void run() {
-            do {
-                if (getSize().width > 0 && getSize().height > 0) {
-                    RenderedImage newZones = segment(image);
-                    if (newZones != null) {
-                        zones = newZones;
-                        repaint();
+            while (running.get()) {
+                try {
+                    SegmentRequest request = requestQueue.poll();
+                    if (request == null) {
+                        Thread.sleep(50); // Brief wait if no work
+                        continue;
+                    }
+
+                    if (getSize().width > 0 && getSize().height > 0) {
+                        RenderedImage processedImage = cropScaleGrayscale(request.visibleRect, request.image);
+                        RenderedImage newZones = segment(processedImage);
+                        if (newZones != null) {
+                            zones = newZones;
+                            SwingUtilities.invokeLater(() -> repaint());
+                        }
                     }
+                } catch (InterruptedException e) {
+                    break;
+                } catch (Exception e) {
+                    e.printStackTrace();
                 }
-            } while (getNextView());
+            }
+        }
+    }
+
+    private static class SegmentRequest {
+        final Rectangle visibleRect;
+        final PlanarImage image;
+
+        SegmentRequest(Rectangle visibleRect, PlanarImage image) {
+            this.visibleRect = visibleRect;
+            this.image = image;
         }
     }
```

Comment:
Replace this class with a record.

```suggestion
    private record SegmentRequest(Rectangle visibleRect, PlanarImage image) { }
```